File: ruby/bob/bob.rb
# frozen_string_literal: true

module Bob
  module_function

  def hey(input)
    if (remark = input.strip).blank?
      return "Fine. Be that way!"
    end

    respond_to_verbal_remark(remark)
  end

  private_class_method def respond_to_verbal_remark(remark)
    case [question?(remark), shouting?(remark)]
    when [true, true]
      "Calm down, I know what I'm doing!"
    when [true, false]
      "Sure."
    when [false, true]
      "Whoa, chill out!"
    else
      "Whatever."
    end
  end

  private_class_method def question?(remark)
    remark.end_with?("?")
  end

  private_class_method def shouting?(remark)
    remark == remark.upcase && remark != remark.downcase
  end
end

File: haskell/bob/src/Bob.hs
{-# LANGUAGE OverloadedStrings #-}

module Bob (responseFor) where

import Data.Function ((&))
import Data.Text (Text)
import qualified Data.Text as Text

responseFor :: String -> String
responseFor input =
  let parsedInput = input & Text.pack & Text.strip
  in
    case parsedInput of
      "" -> "Fine. Be that way!"

      remark -> respondToVerbalRemark remark

-- PRIVATE

respondToVerbalRemark :: Text -> String
respondToVerbalRemark remark =
  let
    isQuestion = Text.isSuffixOf "?" remark

    isShouting = Text.toUpper remark == remark && Text.toLower remark /= remark
  in case (isQuestion, isShouting) of
    (True, True) -> "Calm down, I know what I'm doing!"

    (True, False) -> "Sure."

    (False, True) -> "Whoa, chill out!"

    (False, False) -> "Whatever."
